Adding a preprocessor check for which C library the compiler is using.

JohnTortugo opened the pull request:
This is related to this issue: https://github.com/dotnet/core-eng/issues/2204

The problem was related to the prototype/implementation of the ioctl function in the MUSL library used by the Alpine OS.

The Alpine OS uses a C library implementation called MUSL. This library declares/implements the ioctl function as ioctl(int, int, ...). However, most Linux distros use the Glibc which declares/implements the ioctl function as ioctl(int, unsigned long, ...). Since a parameter was used to tell the compiler to consider as error the conversions from wider types (unsigned long) to smaller ones (int) the compilation was halting.

A preprocessor check was added to check whether the program is being compiled using the Glibc or not.

stephentoub commented:
Thanks, @JohnTortugo.  Does the current code (the result of https://github.com/dotnet/corefx/pull/25687) still result in problems?  If that code compiles ok on Alpine (I expect it would), it's fine to just leave that rather than adding this ifdef'ing.

Separately, I assume the reason we don't see this failure in CI is because we don't build for Alpine in CI?  Is that something we can fix? cc: @janvorli

JohnTortugo commented:
The code worked fine on Alpine. I just refactored the routines to how they were before and added the check.

> Separately, I assume the reason we don't see this failure in CI is because we don't build for Alpine in CI?

I am not sure about that, but I would like to be :-)

stephentoub commented on src/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -2272,48 +2272,43 @@ int32_t SystemNative_Socket(int32_t addressFamily, int32_t socketType, int32_t p
 }
 
-int32_t SystemNative_GetBytesAvailable(intptr_t socket, int32_t* available)
+int32_t SystemNative_GetAtOutOfBandMark(intptr_t socket, int32_t* available)
```
I don't have a strong opinion about whether we take this change or just keep the code as it is and close this PR, with a slight preference for not adding more ifdefs and just closing the PR.  But if we do take the change, we should keep the parameter name as `atMark` rather than `available`.  And the `available` parameter to `GetInt32Ioctl` should also be renamed to `result` or something similar (the `avail` local should also be changed).

JohnTortugo replied:
I'll go with the simpler path of keeping the existing code. I'll close this PR.

janvorli commented:
I have not seen this issue when building corefx on Alpine locally and we build corefx for Alpine 3.6 in the lab twice a day to get official builds and that build succeeds. So I wonder what was the Alpine Linux version for which it was not compiling.
Even if the change is needed, we should not use the `__GLIBC__` symbol to detect whether a function has certain parameters or not, since it is fragile. The right way would be to detect that at build time in configure.cmake mechanism and produce a symbol named e.g. HAVE_IOCTL_2ND_ARG_INT or something along these lines and use it here instead.

JohnTortugo commented:
I am using this Docker image: alpine-3.6-3148f11-20171119021156 from this repo: alpine-3.6-3148f11-20171119021156
